docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,216 @@
+```text
+group: a
+   alert:    <name:"a1" last_evaluation:<seconds:1 > > 
+   recording:<name:"r1" last_evaluation:<seconds:1 > > 
+   recording:<name:"r2" last_evaluation:<seconds:1 > > 
+group: b
+   recording:<name:"r1" last_evaluation:<seconds:1 > > 
+```
+
+Note in the example above how the recording rule `r1` is not deduplicated as it is contained in two different groups.
+
+Scenario 2:
+
+The next level of deduplication is governed by the label/value set of the underlying recording/alerting rule while respecting the replica label.
+For a given conflict, the youngest is preferred. For alerting rules, the youngest firing rule is preferred.
+
+Given the following stream of incoming recording rules:
+```text
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > > 
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > > 
+```
+
+The output becomes:
+```text
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > > 
+```
+
+Given the following stream of incoming alerting rules:
+```text
+group: a
+   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > > 
+group: a
+   alert:<             name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > > 
+```
+
+The output becomes:
+```text
+group: a
+   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > > 
+```
+
+Note how in the above output the firing alerting rule was preferred despite being older.
+
+Scenario 3:
+
+If, under the above conditions a rule is a candidate for deduplication, finally the rule `expr` and `for` fields are being considered for deduplication.
+
+Given the following stream of incoming alerting rules will also result in two independent alerting rules as both the `expr` and `for` fields differ:
+```text
+  - alert: KubeAPIErrorBudgetBurn
+    annotations:
+      message: The API server is burning too much error budget
+    expr: |
+      sum(apiserver_request:burnrate1h) > (14.40 * 0.01000)
+      and
+      sum(apiserver_request:burnrate5m) > (14.40 * 0.01000)
+    for: 2m
+    labels:
+      severity: critical
+  - alert: KubeAPIErrorBudgetBurn
+    annotations:
+      message: The API server is burning too much error budget
+    expr: |
+      sum(apiserver_request:burnrate6h) > (6.00 * 0.01000)
+      and
+      sum(apiserver_request:burnrate30m) > (6.00 * 0.01000)
+    for: 15m
+    labels:
+      severity: critical
+```
+
+### Type and Info
+
+To accommodate for future additional APIs alongside with the core Rules proto changes an enhancement of the protobuf Info and Type structures are suggested.
+
+ The current `StoreType` enum is renamed to `Type`. This retains binary compatibility with older clients:
```

Comment:
Why do we need this? Can this be a different prosposal? as right now the main proposal introduces new Service and doesn't change existing protos or services.

Reply:
yes, I agree, this came up during a pairing session with @bwplotka but I agree we should handle this in a separate proposal.